File: src/main/java/ru/romanow/scc/warehouse/domain/Items.java
package ru.romanow.scc.warehouse.domain;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;
import lombok.Data;
import lombok.experimental.Accessors;

import javax.persistence.*;
import java.util.List;
import java.util.UUID;

@Data
@Accessors(chain = true)
@Entity
@Table(name = "items",
        indexes = {
                @Index(name = "idx_items_uid", columnList = "uid", unique = true),
                @Index(name = "idx_items_name", columnList = "name", unique = true)
        })
public class Items {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(name = "uid", nullable = false, unique = true, updatable = false)
    private UUID uid;

    @Column(name = "name", length = 1024, nullable = false, unique = true)
    private String name;

    @Column(name = "count")
    private Integer count;

    @ManyToMany(mappedBy = "items")
    private List<OrderItems> orderItems;

    public Integer decrementCount() {
        return --count;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Items items = (Items) o;
        return Objects.equal(name, items.name);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("uid", uid)
                .add("name", name)
                .add("count", count)
                .toString();
    }
}

File: src/main/java/ru/romanow/warehouse/domain/Items.kt
package ru.romanow.warehouse.domain

import java.util.*
import javax.persistence.*

@Entity
@Table(
    name = "items",
    indexes = [
        Index(name = "idx_items_uid", columnList = "uid", unique = true),
        Index(name = "idx_items_name", columnList = "name", unique = true)
    ]
)
data class Items(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Int? = null,

    @Column(name = "uid", nullable = false, unique = true, updatable = false)
    var uid: UUID? = null,

    @Column(name = "name", length = 1024, nullable = false, unique = true)
    var name: String? = null,

    @Column(name = "count", nullable = false)
    var count: Int? = null,

    @ManyToMany(mappedBy = "items")
    var orderItems: List<OrderItems>? = null,
) {
    fun decrementCount(): Int {
        count = (count ?: 1) - 1
        return count!!
    }

    override fun toString(): String {
        return "Items(id=$id, uid=$uid, name=$name, count=$count)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Items

        if (uid != other.uid) return false
        if (name != other.name) return false
        if (count != other.count) return false

        return true
    }

    override fun hashCode(): Int {
        var result = uid?.hashCode() ?: 0
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + (count ?: 0)
        return result
    }
}
